core/lib/authorization/capabilities.ts

```diff
@@ -0,0 +1,211 @@
+import type { CommunitiesId, FormsId, PubsId, StagesId, UsersId } from "db/public";
+import { Capabilities } from "db/src/public/Capabilities";
+import { MembershipType } from "db/src/public/MembershipType";
+
+import { db } from "~/kysely/database";
+
+export const pubCapabilities = [
+	Capabilities.movePub,
+	Capabilities.createPub,
+	Capabilities.viewPub,
+	Capabilities.deletePub,
+	Capabilities.updatePubValues,
+	Capabilities.createRelatedPub,
+	Capabilities.editPubWithForm,
+	Capabilities.addPubMember,
+	Capabilities.removePubMember,
+	Capabilities.runAction,
+] as const;
+
+export const communityCapabilities = [
+	Capabilities.createPubField,
+	Capabilities.archivePubField,
+	Capabilities.editPubField,
+	Capabilities.createPubType,
+	Capabilities.editPubType,
+	Capabilities.deletePubType,
+	Capabilities.createStage,
+	Capabilities.addCommunityMember,
+	Capabilities.removeCommunityMember,
+	Capabilities.manageMemberGroups,
+	Capabilities.editCommunity,
+	Capabilities.createForm,
+	Capabilities.createApiToken,
+	Capabilities.revokeApiToken,
+] as const;
+export const stageCapabilities = [
+	Capabilities.viewStage,
+	Capabilities.manageStage,
+	Capabilities.deleteStage,
+	Capabilities.addStageMember,
+	Capabilities.removeStageMember,
+] as const;
+export const formCapabilities = [
+	Capabilities.createPubWithForm,
+	Capabilities.editPubWithForm,
+	Capabilities.addFormMember,
+	Capabilities.removeFormMember,
+	Capabilities.editForm,
+	Capabilities.archiveForm,
+] as const;
+
+type CapabilitiesArg = {
+	[MembershipType.pub]: typeof pubCapabilities;
+	[MembershipType.stage]: typeof stageCapabilities;
+	[MembershipType.community]: typeof communityCapabilities;
+	[MembershipType.form]: typeof formCapabilities;
+};
+
+type Target = PubTarget | StageTarget;
+
+type PubTarget = {
+	type: MembershipType.pub;
+	pubId: PubsId;
+};
+
+type StageTarget = {
+	type: MembershipType.stage;
+	stageId: StagesId;
+};
+
+type CommunityTarget = {
+	type: MembershipType.community;
+	communityId: CommunitiesId;
+};
+
+type FormTarget = {
+	type: MembershipType.form;
+	formId: FormsId;
+};
+
+export const userCan = async <T extends Target>(
+	capability: CapabilitiesArg[T["type"]][number],
+	target: T,
```

Comment:
There's a distinction between capabilities that's not really captured in the database structure. `updatePubValues` is a pub capability because its target is a pub, meaning we need a pubId to determine if the user has that capability. `viewStage` is a stage capability so it requires a stageId. The types here make sure that you'll get a type error if you call `userCan` with a pub capability but pass a stageId or any other invalid combination.

Reply:
There's one very misleading thing here, which is my use of MembershipType to express both the type of membership the user has, and the type of target a certain capability requires. At the moment those are both `pub`, `stage`, `community`, or `form`, but that's kind of incidental! Let me know if that makes sense and/or if it's troubling.

Reply:
I'm not sure I follow—when would they be different?

Reply:
I think you're right, they wouldn't be. If we were to add a new kind of entity and associated capabilities for it but didn't want to add a new membership type, that's when they'd be different. But I don't think that's very likely!